### PR TITLE
feat: Added a generic LRUCache interface and a default implementation

### PR DESCRIPTION
**Pull request by mikechu-optimizely**
## Summary
Added an interface for a LRU Caching mechanism using generics.  Include a default implementation.

## Test plan
Included new tests in the test project

## Issues
OASIS-8407

### CODE REVIEW

File: OptimizelySDK/Odp/ILruCache.cs
Comment (jaeopt) [inline OptimizelySDK/Odp/ILruCache.cs:19]: Let's drop "Lru" since this is generic cache interface not only for LRU.

File: OptimizelySDK/Odp/LruCache.cs
Comment (jaeopt) [inline OptimizelySDK/Odp/LruCache.cs:38]: Let's set to zero if maxSize is negative. For consistency with other SDKs.
```suggestion
            _maxSize = maxSize < 0 ? 0 : maxSize;
```

**Review by jaeopt — APPROVED** [review]
LGTM

**Comment by mikechu-optimizely** [comment]
@jaeopt Would you like for this Pull Request to also wait for @msohailhussain 's approval?

**Comment by jaeopt** [comment]
> @jaeopt Would you like for this Pull Request to also wait for @msohailhussain 's approval?

Yes for all PRs. Unfortunately, I cannot cover csharp lang stuffs :)

File: OptimizelySDK/Odp/LruCache.cs
Comment (msohailhussain) [inline OptimizelySDK/Odp/LruCache.cs:82]: am not a big fan of casting, Is it possible if you can maintain a Queue along with typed Dictionary. This will reduce casting time because we only need ordered dictionary when our queue is full.

File: OptimizelySDK.Tests/OdpTests/LruCacheTest.cs
Comment (msohailhussain) [inline OptimizelySDK.Tests/OdpTests/LruCacheTest.cs:138]: Is this possible if we can return `true` or `false`. @jaeopt a question for you?

File: OptimizelySDK/Odp/LruCache.cs
Comment (msohailhussain) [inline OptimizelySDK/Odp/LruCache.cs:26]: Use TimeSpan.

File: OptimizelySDK/Odp/LruCache.cs
Comment (msohailhussain) [inline OptimizelySDK/Odp/LruCache.cs:36]: I think default values are much better option than having 2 constructors.

**Review by msohailhussain — CHANGES_REQUESTED** [review]
please address

File: OptimizelySDK/Odp/LruCache.cs
Comment (mikechu-optimizely) [reply OptimizelySDK/Odp/LruCache.cs:122]: `ReadCurrentCache` is an unfortunate shortcut for me to read the state ordered dictionary in unit test.  I've removed.

Comment (mikechu-optimizely) [reply OptimizelySDK/Odp/LruCache.cs:122]: JetBrains noted, "The field is sometimes used inside synchronized block and sometimes used without synchronization" which seemed logical.

File: OptimizelySDK/Odp/LruCache.cs
Comment (mikechu-optimizely) [reply OptimizelySDK/Odp/LruCache.cs]: What's your opinion of using pattern matching instead of the hard cast?

File: OptimizelySDK/Odp/LruCache.cs
Comment (mikechu-optimizely) [reply OptimizelySDK/Odp/LruCache.cs:31]: Can I put the const 600 in the constructor?

File: OptimizelySDK/Odp/LruCache.cs
Comment (mikechu-optimizely) [reply OptimizelySDK/Odp/LruCache.cs:28]: Can you clarify your thoughts here?  DM me if it's easier.

**Comment by mikechu-optimizely** [comment]
@msohailhussain Can you take a look at this Pull Request and the suggested LRU implementation when you have a chance?

File: .editorconfig
Comment (msohailhussain) [inline .editorconfig:16]: thanks

File: OptimizelySDK/Odp/LruCache.cs
Comment (msohailhussain) [inline OptimizelySDK/Odp/LruCache.cs:24]: document this class please.

File: OptimizelySDK/Odp/LruCache.cs
Comment (msohailhussain) [inline OptimizelySDK/Odp/LruCache.cs:35]: what should be the default? `null` is not default. Let's talk offline.

File: OptimizelySDK/Odp/LruCache.cs
Comment (msohailhussain) [inline OptimizelySDK/Odp/LruCache.cs:45]: please use DEFAULT_MAX_SIZE in the args. 
Don't need CACHE_DISABLED; we can check if it's less than 0.

File: OptimizelySDK/Odp/LruCache.cs
Comment (msohailhussain) [inline OptimizelySDK/Odp/LruCache.cs:58]: same here.

File: OptimizelySDK/Odp/LruCache.cs
Comment (msohailhussain) [inline OptimizelySDK/Odp/LruCache.cs:82]: We should have logging here.

File: OptimizelySDK/Odp/LruCache.cs
Comment (msohailhussain) [inline OptimizelySDK/Odp/LruCache.cs:74]: Am just curious can't we just take LinkedListNode<ItemWrapper> as value?

File: OptimizelySDK/Odp/LruCache.cs
Comment (msohailhussain) [inline OptimizelySDK/Odp/LruCache.cs:137]: should it be in `else` statement?
If this is not caught by unit test, please add a unit test for it.

**Review by msohailhussain — CHANGES_REQUESTED** [review]
have some questions.

File: OptimizelySDK/Odp/LruCache.cs
Comment (mikechu-optimizely) [reply OptimizelySDK/Odp/LruCache.cs]: Yes, let's talk.  I also need help with the default TimeSpan since it can't be a compile-time const.

Comment (mikechu-optimizely) [reply OptimizelySDK/Odp/LruCache.cs]: I've pushed a new solution for discussion.

File: OptimizelySDK/Odp/LruCache.cs
Comment (mikechu-optimizely) [reply OptimizelySDK/Odp/LruCache.cs:162]: I think the `else` is redundant.  The item was expired and needed to be removed from both the index (`LinkedList`) and the cache itself. 

The unit test `ShouldHandleWhenItemsExpire` should be covering this case.

File: OptimizelySDK/Odp/LruCache.cs
Comment (mikechu-optimizely) [reply OptimizelySDK/Odp/LruCache.cs:89]: I think the idea is to pop off both the node from the Tuple and put it in the `_list` & back in the `_cache` along with the `item`  see linw 150ish.

Let's talk through this too.

**Comment by mikechu-optimizely** [comment]
@msohailhussain This is ready for another review after our conversation.

**Review by msohailhussain — APPROVED** [review]
lgtm

**Pull request merged**
